src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java

```diff
@@ -0,0 +1,249 @@
+                    AsynchronousFileChannel.open(path, StandardOpenOption.READ);
+            ByteBuffer buffer = ByteBuffer.allocate((int) file.length());
+            afc.read(
+                    buffer,
+                    0,
+                    buffer,
+                    new CompletionHandler<Integer, ByteBuffer>() {
+                        @Override
+                        public void completed(Integer result, ByteBuffer attachment) {
+                            if (attachment.hasArray()) {
+                                int offset = attachment.arrayOffset();
+                                byte[] data = attachment.array();
+                                cb.onGetComplete(entry.toCacheEntry(data));
+                            }
+                        }
+
+                        @Override
+                        public void failed(Throwable exc, ByteBuffer attachment) {
+                            cb.onGetComplete(null);
+                        }
+                    });
+        } catch (Exception e) {
+            cb.onGetComplete(null);
+        }
+    }
+
+    @Override
+    public void put(String key, Entry entry, OnPutCompleteCallback callback) {
+        // TODO (sphill99): Implement
+    }
+
+    /**
+     * Initializes the DiskBasedAsyncCache by scanning for all files currently in the specified root
+     * directory. Creates the root directory if necessary.
+     */
+    @Override
+    public synchronized void initialize() {
+        // TODO (sphill99): Implement
+    }
+
+    /**
+     * Invalidates an entry in the cache.
+     *
+     * @param key Cache key
+     * @param fullExpire True to fully expire the entry, false to soft expire
+     */
+    @Override
+    public synchronized void invalidate(String key, boolean fullExpire) {
+        // TODO (sphill99): Implement
+    }
+
+    /** Removes the specified key from the cache if it exists. */
+    @Override
+    public synchronized void remove(String key) {
+        // TODO (sphill99): Implement
+    }
+
+    /**
+     * Creates a pseudo-unique filename for the specified cache key.
+     *
+     * @param key The key to generate a file name for.
+     * @return A pseudo-unique filename.
+     */
+    private String getFilenameForKey(String key) {
+        int firstHalfLength = key.length() / 2;
+        String localFilename = String.valueOf(key.substring(0, firstHalfLength).hashCode());
+        localFilename += String.valueOf(key.substring(firstHalfLength).hashCode());
+        return localFilename;
+    }
+
+    /** Returns a file object for the given cache key. */
+    public File getFileForKey(String key) {
+        return new File(mRootDirectorySupplier.get(), getFilenameForKey(key));
+    }
+
+    /** Represents a supplier for {@link File}s. */
+    public interface FileSupplier {
+        File get();
+    }
+
+    /** Handles holding onto the cache headers for an entry. */
+    @VisibleForTesting
+    static class CacheHeader {
```

Comment:
Could `CacheHeader` be shared between this class and `DiskBaseCache`? The implementations don't look the same at this time, but just wondering what's the long term plan and if it would make sense to extract the class to a top-level class so it can be shared.

Reply:
Yeah, I was thinking about making another abstract class that these two extend. This would allow us to not repeat the shared methods/class. Currently, I think the write and read methods would be implemented differently, so I don't know how that'd work. But I think that's definitely better to try to do that.